paygd: Write logs to a separate file

jprvita opened the pull request:
In systems with fragile storage (ex., SSD or eMMC) the journal is only
kept in RAM as does not persist over reboot. In such systems we still
want logs from eos-paygd to persist, without having to write all system
logs to persistent storage.

The solution for that is to create a separate log file for eos-paygd, in
addition to forwarding logs to the journal.

https://phabricator.endlessm.com/T33220

wjt commented on eos-paygd/main.c:
```diff
@@ -240,6 +295,9 @@ main (int   argc,
   GOptionContext *context;
   gboolean enforcing_mode = TRUE;
 
+  log_io = g_io_channel_new_file (LOG_FILE_NAME, "a", NULL);
```
Is the real `/var` accessible when `eos-payg` starts up?

You are not handling failure from this function, indicated by it returning `NULL`. https://docs.gtk.org/glib/ctor.IOChannel.new_file.html

You could write:

```
  log_io = g_io_channel_new_file (LOG_FILE_NAME, "a", &error);
  if (log_io != NULL)
    {
      g_log_set_writer_func (log_writer, log_io, NULL);
    }
  else
    {
      g_warning ("Failed to open %s: %s", LOG_FILE_NAME, error->message);
      g_clear_error (&error);
    }
```

jprvita replied:
> Is the real `/var` accessible when `eos-payg` starts up?
> 

Hum, you are right, we need to wait until pivoting to the real root, which means we'll lose some startup messages (I). That is fine IMO, as these will still be in the journal and are either fatal errors or redundant debug info: nothing that would affect an analysis of credit consumption.

wjt commented on eos-paygd/main.c:
```diff
@@ -268,6 +271,76 @@ test_and_update_securitylevel (void)
+  /* Use /bin/mkdir instead of mkdir() to ensure the mode is unaffected by
+   * the process's umask.
+   */
+  system_ret = system ("/bin/mkdir -p --mode=755 " LOGFILE_DIRNAME);
```
Can we create the directory with `systemd-tmpfiles` instead?

wjt replied:
Or, looks like we can use `LogsDirectory=eos-paygd` in the unit file to cause systemd to create it before launching eos-paygd.

jprvita replied:
I don't think so, as `systemd-tmpfiles` would not have run before paygd starts, and `LogsDirectory=` would add a `RequiresMountsFor=/var` dependency which I believe would either delay paygd's startup or create an impossible transaction.

I'm using `mkdir` following how `/var/lib/eos-payg` is created a few lines below.

wjt commented on eos-paygd/main.c:
```diff
@@ -271,6 +272,14 @@ test_and_update_securitylevel (void)
   return TRUE;
 }
 
+static gint
+sort_func (gconstpointer p1, gconstpointer p2)
+{
+  const gchar *a = (const char *) p1;
+  const gchar *b = (const char *) p2;
+  return g_strcmp0 (a, b);
+}
```
You don't need this, you can just pass `g_strcmp0` as the compare_func.

wjt replied:
Cast it: `(GCompareFunc) g_strcmp0`

jprvita replied:
:brain:

wjt reviewed (changes requested):
I confess to not having read the cleanup logic in detail.

I wonder if we could instead use systemd-tmpfile to log files older than a certain age; and rather than having logic to rename old files, always name the log file after the current timestamp. This would be a lot less code!

jprvita commented on eos-paygd/main.c:
```diff
@@ -240,6 +295,9 @@ main (int   argc,
   gboolean enforcing_mode = TRUE;
 
+  log_io = g_io_channel_new_file (LOG_FILE_NAME, "a", NULL);
+  g_log_set_writer_func (log_writer, log_io, NULL);
```
Hum, I thought this would try to free `user_data` after each call to the callback. Are you sure the way you describe is how it works?

I supposed I could also close the io channel on the last line before returning from main as well.

jprvita commented on eos-paygd/main.c:
```diff
@@ -35,8 +35,62 @@
+            gpointer user_data)
+{
+  g_autofree gchar *out = NULL;
+  GIOChannel *log_io = (GIOChannel *) user_data;
```
Thanks!

wjt replied:
Or maybe not, this is not a GObject!

jprvita commented:
Addressed feedback provided in this PR, plus @wjt's suggestion to split logs as one file per day then use systemd-tmpfiles to cleanup files older then a certain age (I have chosen 90 days).

(I haven't been able to test the 90 days removal yet)

jprvita commented:
Ah, one minor annoyance that I noticed when testing the build changes:
```
tmpfilesdir = get_option('tmpfilesdir')
if tmpfilesdir == ''
  tmpfilesdir = systemd.get_pkgconfig_variable('tmpfilesdir')
endif
```
This ends up installing the tmpfile in the system's `/usr` instead of under the prefix. The same is true for the unit files. It does not matter unless you build and install eos-payg locally for some reason. It would be nice to fix if someone happens to know the right way from the top of their head, but it is not worth putting any actual effort into it IMO.

wjt commented on eos-paygd/meson.build:
```diff
@@ -36,6 +36,10 @@ systemdsystemunitdir = get_option('systemdsystemunitdir')
 if systemdsystemunitdir == ''
   systemdsystemunitdir = systemd.get_pkgconfig_variable('systemdsystemunitdir')
 endif
+tmpfilesdir = get_option('tmpfilesdir')
+if tmpfilesdir == ''
+  tmpfilesdir = systemd.get_pkgconfig_variable('tmpfilesdir')
+endif
```
Try this:

```suggestion
tmpfilesdir = get_option('tmpfilesdir')
if tmpfilesdir == ''
  tmpfilesdir = systemd.get_pkgconfig_variable('tmpfilesdir', define_variable: ['prefix', prefix])
endif
```

Found at https://gitlab.freedesktop.org/pipewire/pipewire/-/merge_requests/235/diffs

wjt commented on eos-paygd/main.c:
```diff
@@ -35,10 +35,51 @@
+   * ex., eos-paygd-20220418.log
+   */
+  now = g_date_time_new_now_local ();
+  tstamp = g_date_time_format (now, "%Y%m%d");
```
Why not `%F`, the ISO-8601 date format (`%Y-%m-%d`)?

jprvita replied:
Just personal preference of making it a number and shorter. I think `eos-paygd-20220418.log` reads better than `eos-paygd-2022-04-18.log` or `eos-paygd_2022-04-18.log` (the latter being slightly better, but I still prefer the current format.

wjt commented on eos-paygd/main.c:
```diff
@@ -583,11 +574,22 @@ main (int   argc,
    */
-  if (open_log_file (&log_io))
-    g_log_set_writer_func (log_writer, log_io, NULL);
+  system_ret = system ("/bin/mkdir -p --mode=755 " LOGFILE_DIRNAME);
```
Does this run before or after the real `/var` is available?

jprvita replied:
Yes, `/var` is mounted by `ostree-prepare-root.service`, ordered `Before=initrd-root-fs.target`, whereas `eos-paygd-2.service` is ordered `After=initrd-root-fs.target`.

At this point in the code paygd has chroot'ed itself to `/sysroot` and was able to connect to D-Bus's system bus through a busy loop, so we now the real root is in a good state.

I think we can actually create the logdir earlier by pointing it to `/sysroot/var` when running from the initramfs, but I need to try it out.

jprvita commented on eos-paygd/main.c:
```diff
@@ -35,8 +35,95 @@
+  now = g_date_time_new_now_local ();
+  tstamp = g_date_time_format (now, "%Y%m%d");
+  log_file_name = g_strconcat (LOGFILE_DIRNAME, "/", LOGFILE_BASENAME, "-",
+                               tstamp, "." LOGFILE_EXT, NULL);
```
```suggestion
                               tstamp, ".", LOGFILE_EXT, NULL);
```
Oops! (Still works due to the magical string literal concatenation performed by cpp).

jprvita commented:
Addressed a couple of the smaller problems highlighted in the comments above on the original commits, and added an extra commit that tries to start logging to this separate file as early as possible. This last part is still untested but ready for initial feedback.

jprvita commented:
> added an extra commit that tries to start logging to this separate file as early as possible. This last part is still untested but ready for initial feedback.

Ok, this did not work right away on my initial draft, and I don't think it is worth the debugging effort for the benefit of getting a few extra early debug messages. I'm reverting this extra commit, so the only changes from the last round were the fixes for comments in the PR (both mine and from Will).

jprvita commented:
> > added an extra commit that tries to start logging to this separate file as early as possible. This last part is still untested but ready for initial feedback.
> 
> Ok, this did not work right away on my initial draft, and I don't think it is worth the debugging effort for the benefit of getting a few extra early debug messages. I'm reverting this extra commit, so the only changes from the last round were the fixes for comments in the PR (both mine and from Will).

This is the comparison between the previously discussed revision and the current one: https://github.com/endlessm/eos-payg/compare/8e8c66583b374e0dc7da642cb44374d307b01408..T33220

jprvita commented:
> (I haven't been able to test the 90 days removal yet)

I've tested this by moving the clock fwd in the BIOS then waiting 15 minutes after boot for systemd-tmpfiles-clean.service to run, and it removed files older than 90 days, as expected.

wjt commented on eos-paygd/main.c:
```diff
@@ -35,8 +35,92 @@
+  log_file = g_file_new_for_path (log_file_name);
+
+  *log_io = g_io_channel_new_file (log_file_name, "a", NULL);
+  if (!log_io)
```
You mean `if (!*log_io)` here. This function as written will either dereference a null pointer, or return TRUE unconditionally.

You could choose to instead return the `GIOChannel *` rather than `gboolean` and have no out parameter, which would make this bug harder to write.

jprvita replied:
Oh, good catch, thanks! Yeah, I usually don't like out parameters because it is easy to overlook the dereferencing. There was something on the previous versions that made me think it was better that way, but with the way the code is right now I agree ditching the out parameter entirely.

pwithnall commented on eos-paygd/main.c:
```diff
@@ -35,8 +35,90 @@
+   */
+  now = g_date_time_new_now_local ();
+  tstamp = g_date_time_format (now, "%Y%m%d");
+  log_file_name = g_strconcat (LOGFILE_DIRNAME, "/", LOGFILE_BASENAME, "-",
```
Technically this should be `g_build_filename()` rather than `g_strconcat()`, for portability to other platforms and for the case where the file system encoding is not UTF-8 (which is possible but vanishingly rare). Just an fyi, though, this isn’t worth changing now. The filename would have to be constructed separately (using `g_strconcat()`) before building the path.

pwithnall replied:
Actually, the return value from `g_strconcat()` is leaked here, so this does need some tweaking.

pwithnall commented on eos-paygd/main.c:
```diff
@@ -35,8 +35,90 @@
+  return g_io_channel_new_file (log_file_name, "a", NULL);
+}
+
+/* we can't use g_log_writer_default_would_drop until glib 2.68, which will be
```
Nitpick: I would have used (pseudocode) `#if GLIB_CHECK_VERSION(2, 68, 0) { return g_log_writer_default_would_drop (…) } #else { … simplified fallback … }` so that it’s easier to find and drop the fallback code in future (by grepping for `GLIB_CHECK_VERSION`). Again, just an fyi, not worth changing now.

pwithnall commented on eos-paygd/main.c:
```diff
@@ -35,8 +35,90 @@
+  if (should_drop_message (log_level))
+    return G_LOG_WRITER_HANDLED;
+
+  log_io = open_log_file ();
```
Should this be opening the log file, writing to it, and closing it again every time a message is logged? That seems costly.

Would it be better to keep a persistent handle to the open log file?

jprvita replied:
Re-opening the file every time is the way we designed cheap log rotation: we can't use `logrotate` because eos-payg can't receive signals in "phase-4 mode", and while there are other design alternatives (use logrotate and monitor the fd for renames with inotify, monitor the date or file size and close + re-open at that point), they are all more complicated and this seems good enough for what we need while remaining relatively pretty simple.

This designed choice is mentioned in the commit message of https://github.com/endlessm/eos-payg/commit/4a98f292cce82d260423aef614f7aede580a58e1, but it seems like we should also add it as a comment in the code.

pwithnall replied:
:+1:

pwithnall commented on eos-paygd/main.c:
```diff
@@ -487,6 +569,23 @@ main (int   argc,
+  /* Use /bin/mkdir instead of mkdir() to ensure the mode is unaffected by
+   * the process's umask.
+   */
+  system_ret = system ("/bin/mkdir -p --mode=755 " LOGFILE_DIRNAME);
```
Is this not done by the systemd-tmpfiles snippet? Sorry if that’s already been covered in a discussion; I only skimmed through them.

*Edit*: Oh, I see it was [discussed here](https://github.com/endlessm/eos-payg/pull/104#discussion_r854633924). Perhaps the details about startup ordering and dependencies on `RequiresMountsFor=` are worth a code comment here?

pwithnall reviewed (commented):
Here are a few post-hoc review comments. Sorry for nitpicking.

I’ve tried to address them in https://github.com/endlessm/eos-payg/pull/105. I have not addressed the comment about reopening the log file for each message being logged, as that seemed like it needed a bit more discussion before being changed (if at all).
